ynr/apps/sopn_parsing/helpers/extract_pages.py

```diff
@@ -27,36 +28,45 @@ def extract_pages_for_ballot(ballot, manual_upload=False):
             f"Warning! HTML file types are not supported; trying to convert {ballot.sopn.uploaded_file.path} to pdf"
         )
         html_file_path = ballot.sopn.uploaded_file.path
-        pdf_file_path = ballot.sopn.uploaded_file.path + ".pdf"
-        # Replacing the extention with .pdf also works and
-        # negates the need for pdfkit but not sure
-        # how reliable it is
-        # pdf_file = ballot.sopn.uploaded_file.path.replace(".html", ".pdf")
-        config = pdfkit.configuration(wkhtmltopdf="/usr/local/bin/wkhtmltopdf")
-        options = {"enable-local-file-access": None}
-
-        pdfkit.from_file(
-            html_file_path,
-            pdf_file_path,
-            configuration=config,
-            verbose=True,
-            options=options,
+        html_file_name = ballot.sopn.uploaded_file.name
+        pdf_file_path = html_file_path + ".pdf"
+
+        pypandoc.convert_file(
+            html_file_path, to="pdf", outputfile=pdf_file_path
+        )
+
+        pdf_file_path.replace(".html", "")
+        pdf_file_name = html_file_name.replace(".html", "")
+
+        # get the html file
+        html_document = OfficialDocument.objects.filter(
+            source_url=ballot.sopn.source_url
+        ).order_by("-uploaded_file")[0]
+
+        # TO DO: get the pdf file??
+
+        # Create a new OfficialDocument object with the same source_url but with
+        # the new pdf file.
+        new_pdf_official_document = OfficialDocument(
+            ballot=ballot,
+            source_url=ballot.sopn.source_url,
+            document_type=OfficialDocument.NOMINATION_PAPER,
         )
 
-        # save the pdf file as the sopn document uploaded file
-        # ballot.sopn.uploaded_file.save(
-        #     ballot.sopn.uploaded_file.name,
-        #     File(open(pdf_file_path, "rb")),
-        #     save=True,
-        # )
-        # ballot.sopn.save()
+        # save the uploaded file to the new pdf doc
+        # TO DO: How can I get the content file object such as...
+        # file_object = ContentFile(content=file_response.content)
 
```

Comment:
@michaeljcollinsuk I'm not exactly sure how to save the newly created `OfficialDocument` to the current `ballot.sopn` and therefore overwrite the original upload. Am I on the right path here?

Reply:
So if I understand correctly, `pypandoc` is writing a new file to disk at the path of `pdf_file_path`? You will then need to use that pdf file to create an instance of Django's `File` before it can be saved as the `uploaded_file` of the `OfficialDocument` instance. That is because `OfficialDocument.uploaded_file` is a `FileField`. It's worth reading through the django documentation for `FileField` and in particular for the [save method](https://docs.djangoproject.com/en/4.0/ref/models/fields/#django.db.models.fields.files.FieldFile.save).

I dont think you need to create a new OfficialDocument, you just want to change the `uploaded_file` that is stored on the instance. So after you have converted to a PDF, I think you will want something along the lines of:
```
with open(pdf_file_path) as file:
    ballot.sopn.uploaded_file.save(
        name=pdf_file_name,
        content=file
    )
```
Although I haven't tried this, but hopefully points in right direction!